pulp_2to3_migration/app/plugin/rpm/migrator.py

```diff
@@ -60,5 +87,95 @@ class RpmMigrator(Pulp2to3PluginMigrator):
         Migrate pre-migrated Pulp 2 RPM plugin content.
         """
         first_stage = ContentMigrationFirstStage(cls)
-        dm = DeclarativeContentMigration(first_stage=first_stage)
+        dm = RPMDeclarativeContentMigration(first_stage=first_stage)
         await dm.create()
+
+
+class RPMDeclarativeContentMigration(DeclarativeContentMigration):
+    """
+    A pipeline that migrates pre-migrated Pulp 2 rpm content into Pulp 3.
+    """
+
+    def pipeline_stages(self):
+        """
+        Build a list of stages.
+
+        This defines the "architecture" of the content migration to Pulp 3.
+
+        Returns:
+            list: List of :class:`~pulpcore.plugin.stages.Stage` instances
+
+        """
+        pipeline = [
+            self.first_stage,
+            QueryExistingArtifacts(),
+            ArtifactSaver(),
+            QueryExistingContents(),
+            ContentSaver(),
+            InterrelateContent(),
+            RelatePulp2to3Content(),
+        ]
+
+        return pipeline
+
+
+class InterrelateContent(Stage):
+    """
+    Stage for relating Content to other Content.
+    """
+
+    async def run(self):
+        """
+        Relate each item in the input queue to objects specified on the DeclarativeContent.
+        """
+        async for batch in self.batches():
+            modulemd_packages_batch = []
+            with transaction.atomic():
+                for dc in batch:
+                    if type(dc.content) == pulp3_models.Modulemd:
+                        thru = self.relate_package_to_module(dc)
+                        modulemd_packages_batch.extend(thru)
+
+                ModulemdPackages = pulp3_models.Modulemd.packages.through
+                ModulemdPackages.objects.bulk_create(objs=modulemd_packages_batch,
+                                                     ignore_conflicts=True,
+                                                     batch_size=1000)
+
+            for dc in batch:
+                await self.put(dc)
+
+    def relate_package_to_module(self, dc):
+        """
+        Relate a Package to a Module.
+
+        Args:
+            dc (pulpcore.plugin.stages.DeclarativeContent): dc for a Module
+        """
+        ModulemdPackages = pulp3_models.Modulemd.packages.through
+        artifacts_list = dc.content.artifacts
+        # find rpm by nevra
+        # We are relying on the order of the processed DC
+        # RPMs should have passed through ContentSaver stage already
+        pq = Q()
+        for artifact in artifacts_list:
+            nevra = package_utils.nevra(artifact)
+            pq |= Q(
+                name=nevra[0],
+                epoch=nevra[1],
+                version=nevra[2],
+                release=nevra[3],
+                arch=nevra[4],
+                is_modular=True)
+        packages_list = []
+        if pq:
+            packages_list = pulp3_models.Package.objects.filter(pq).only('pk')
```

Comment:
I'm not sure if it's better performance-wise to look for modular RPMs first and search nevra then.
E.g. 
```
modular_qs = pulp3_models.Package.objects.filter(is_modular=True)
packages = pulp3_models.Package.objects.filter(pq).only('pk')
```

@dralley , any thoughts?

Reply:
I usually try to make the queries as specific as I can.